// testcafe-hammerhead/lib/processing/resources/script.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _resourceProcessorBase = _interopRequireDefault(require("./resource-processor-base"));

var _lruCache = _interopRequireDefault(require("lru-cache"));

var _script = require("../script");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class ScriptResourceProcessor extends _resourceProcessorBase.default {
  constructor() {
    super();
    this.jsCache = new _lruCache.default({
      // NOTE: Max cache size is 50 MBytes.
      max: 50 * 1024 * 1024,
      length: function length(n) {
        // NOTE: 1 char ~ 1 byte.
        return n.length;
      }
    });
  }

  processResource(script) {
    if (!script) return script;
    let processedScript = this.jsCache.get(script);

    if (!processedScript) {
      processedScript = (0, _script.processScript)(script, true);
      this.jsCache.set(script, processedScript);
    }

    return processedScript;
  }

  shouldProcessResource(ctx) {
    return ctx.contentInfo.isScript;
  }

}

var _default = new ScriptResourceProcessor();

exports.default = _default;
module.exports = exports.default;
